src/tracker/proto.rs

```diff
@@ -0,0 +1,162 @@
+// use ed25519_dalek::{Signature, Signer, VerifyingKey};
+
+use std::fmt;
+
+use super::*;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Message {
+    pub version: u16,
+    pub from: PeerId,
+    pub payload: Payload,
+}
+
+impl Message {
+    pub fn new(from: PeerId, payload: Payload) -> Self {
+        Self {
+            version: VERSION,
+            from,
+            payload,
+        }
+    }
+    pub fn error(from: PeerId, request_id: Option<u64>, error: RpcError) -> Self {
+        Message::new(from, Payload::Error(ErrorMessage { request_id, error }))
+    }
+    pub fn request(from: PeerId, request: Request) -> Self {
+        Message::new(from, Payload::Request(request))
+    }
+    pub fn response(from: PeerId, response: Response) -> Self {
+        Message::new(from, Payload::Response(response))
+    }
+
+    pub fn as_response(&self) -> Option<&Response> {
+        match &self.payload {
+            Payload::Response(response) => Some(response),
+            _ => None,
+        }
+    }
+    pub fn into_response(self) -> Option<Response> {
+        match self.payload {
+            Payload::Response(response) => Some(response),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Payload {
+    Request(Request),
+    Response(Response),
+    Error(ErrorMessage),
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ErrorMessage {
+    request_id: Option<u64>,
+    error: RpcError,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum RpcError {
+    NeedsProtocolVersion(u16),
+    InvalidSignature,
+    UnsupportedOperation,
+}
+impl fmt::Display for RpcError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl std::error::Error for RpcError {}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Request {
+    pub request_id: u64,
```

Comment:
do you need request IDs?  They're a bit of state the client and server needs to manage and could easily get wrong.  Instead you could use the QUIC stream ID for this if you use a stream per request/response.

Reply:
The protocol should be independent of quic and might be used over other transports, so relying on QUIC stream IDs  won't be possible.